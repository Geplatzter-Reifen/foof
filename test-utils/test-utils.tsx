// https://testing-library.com/docs/react-testing-library/setup/#add-custom-queries
import React from "react";
import { render, RenderOptions } from "@testing-library/react-native";
import { FontAwesomeIconsPack } from "@/components/Font/fontAwesome";
import { mapping } from "@eva-design/eva";
import {
  IconRegistry,
  ApplicationProvider,
  ModalService,
} from "@ui-kitten/components";
import * as eva from "@eva-design/eva";
import { foofDarkTheme, foofLightTheme } from "@/constants/custom-theme";
import { library } from "@fortawesome/fontawesome-svg-core";
import { far } from "@fortawesome/free-regular-svg-icons";
import { fas } from "@fortawesome/free-solid-svg-icons";
import { fab } from "@fortawesome/free-brands-svg-icons";

// Iconpacks laden
library.add(far, fas, fab);

// ModalService wird verwendet, um die Modal-Positionierung zu steuern
ModalService.setShouldUseTopInsets = true;

type Props = {
  children: React.ReactNode;
  use_dark_theme?: boolean;
};
// Der Wrapper wird verwendet, um die Provider zu umschließen, die für die Komponenten notwendig sind
const AllTheProviders = ({ children, use_dark_theme = true }: Props) => {
  const theme = use_dark_theme
    ? { ...eva.dark, ...foofDarkTheme }
    : { ...eva.light, ...foofLightTheme };
  return (
    <>
      <IconRegistry icons={FontAwesomeIconsPack} />
      <ApplicationProvider {...eva} customMapping={mapping} theme={theme}>
        {children}
      </ApplicationProvider>
    </>
  );
};

// Erstellt einen Custom-Render, der die Provider um die Komponenten wickelt
const customRender = (
  component: React.ReactElement,
  options?: Omit<RenderOptions, "wrapper">,
) => {
  return render(component, { wrapper: AllTheProviders, ...options });
};

// Exportiert alle Funktionen der Testing Library, weswegen man diese dann auch von dieser Datei importieren kann
export * from "@testing-library/react-native";
export { customRender as render };
