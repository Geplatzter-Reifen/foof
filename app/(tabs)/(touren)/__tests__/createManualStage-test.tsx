import React from "react";
import { Alert } from "react-native";
import { render, fireEvent, waitFor, act } from "@/test-utils/test-utils";
import CreateManualStage from "../createManualStage";

// Mock Router and Navigation
const mockedNavigate = jest.fn();
const mockedGoBack = jest.fn();

jest.mock("@rnmapbox/maps", () => ({
  MapView: ({ children }: { children?: React.ReactNode }) => (
    <div>{children}</div>
  ),
  Marker: () => <div>Mocked Marker</div>,
  Camera: () => <div>Mocked Camera</div>,
  locationManager: {
    start: jest.fn(),
    stop: jest.fn(),
  },
}));
jest.mock("expo-router", () => ({
  useLocalSearchParams: () => ({ tourId: "123" }),
  useRouter: () => ({
    back: mockedGoBack,
    navigate: mockedNavigate,
    push: jest.fn(),
  }),
  useNavigation: () => ({
    navigate: mockedNavigate,
    goBack: mockedGoBack,
    setOptions: jest.fn(),
  }),
}));

// Mock Alert
jest.spyOn(Alert, "alert");

describe("CreateManualStage Component", () => {
  beforeEach(() => {
    jest.useFakeTimers();
    jest.setSystemTime(new Date("2024-01-01T13:00:00Z")); // Fixed UTC time
  });

  afterEach(() => {
    jest.useRealTimers(); // Restore timers
  });
  test("matches snapshot", async () => {
    const view = render(<CreateManualStage />);
    expect(view).toMatchSnapshot();
  });

  test("displays coordinates input when compass button is pressed", async () => {
    const view = render(<CreateManualStage />);

    const compassButton = view.getByTestId("compass");
    await act(async () => {
      fireEvent.press(compassButton);
    });

    await waitFor(() => {
      expect(view.getByText("Start")).toBeTruthy();
      expect(view.getByText("Ende")).toBeTruthy();
      expect(view.getByText("ERSTELLEN")).toBeTruthy();
      expect(view.getByText("ABBRECHEN")).toBeTruthy();
    });
  });

  test("calls back function when cancel button is pressed", async () => {
    const view = render(<CreateManualStage />);
    const cancelButton = view.getByText("ABBRECHEN");

    await act(async () => {
      fireEvent.press(cancelButton);
    });

    expect(mockedGoBack).toHaveBeenCalledTimes(1);
  });

  test("shows alert when dates are invalid", async () => {
    const { getByText } = render(<CreateManualStage />);
    const createButton = getByText("ERSTELLEN");

    await act(async () => {
      fireEvent.press(createButton);
    });

    await waitFor(() => {
      expect(Alert.alert).toHaveBeenCalledWith(
        "Ung√ºltige Eingabe",
        "Der Startzeitpunkt muss vor dem Endzeitpunkt liegen.",
      );
    });
  });
});
