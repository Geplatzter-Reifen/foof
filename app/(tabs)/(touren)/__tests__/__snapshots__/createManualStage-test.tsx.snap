// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateManualStage Component matches snapshot 1`] = `
<View
  appearance="default"
  level="2"
  style={
    [
      {
        "backgroundColor": "#eaeaea",
      },
      {
        "alignItems": "stretch",
        "flex": 1,
        "paddingVertical": 5,
      },
    ]
  }
>
  <View
    onSelect={[Function]}
    selectedIndex={0}
    style={
      [
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginHorizontal": 5,
        },
        undefined,
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#FF6E14",
          "borderColor": "#FF6E14",
          "borderRadius": 4,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 5,
          "minHeight": 40,
          "minWidth": 40,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 12,
        }
      }
      testID="map-pin"
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={25}
        bbWidth={25}
        focusable={false}
        height={25}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 10,
            },
            {
              "flex": 0,
              "height": 25,
              "width": 25,
            },
          ]
        }
        vbHeight={512}
        vbWidth={320}
        width={25}
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          <RNSVGPath
            d="M16 144a144 144 0 1 1 288 0A144 144 0 1 1 16 144zM160 80c8.8 0 16-7.2 16-16s-7.2-16-16-16c-53 0-96 43-96 96c0 8.8 7.2 16 16 16s16-7.2 16-16c0-35.3 28.7-64 64-64zM128 480l0-162.9c10.4 1.9 21.1 2.9 32 2.9s21.6-1 32-2.9L192 480c0 17.7-14.3 32-32 32s-32-14.3-32-32z"
            fill={
              {
                "payload": 4294967295,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#b8b8b8",
          "borderColor": "#b8b8b8",
          "borderRadius": 4,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 5,
          "minHeight": 40,
          "minWidth": 40,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 12,
        }
      }
      testID="compass"
    >
      <RNSVGSvgView
        align="xMidYMid"
        bbHeight={25}
        bbWidth={25}
        focusable={false}
        height={25}
        meetOrSlice={0}
        minX={0}
        minY={0}
        style={
          [
            {
              "backgroundColor": "transparent",
              "borderWidth": 0,
            },
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 10,
            },
            {
              "flex": 0,
              "height": 25,
              "width": 25,
            },
          ]
        }
        vbHeight={512}
        vbWidth={512}
        width={25}
      >
        <RNSVGGroup
          fill={
            {
              "payload": 4278190080,
              "type": 0,
            }
          }
          font={
            {
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
            }
          }
        >
          <RNSVGPath
            d="M256 512A256 256 0 1 0 256 0a256 256 0 1 0 0 512zm50.7-186.9L162.4 380.6c-19.4 7.5-38.5-11.6-31-31l55.5-144.3c3.3-8.5 9.9-15.1 18.4-18.4l144.3-55.5c19.4-7.5 38.5 11.6 31 31L325.1 306.7c-3.2 8.5-9.9 15.1-18.4 18.4zM288 256a32 32 0 1 0 -64 0 32 32 0 1 0 64 0z"
            fill={
              {
                "payload": 4280295456,
                "type": 0,
              }
            }
            propList={
              [
                "fill",
              ]
            }
          />
        </RNSVGGroup>
      </RNSVGSvgView>
    </View>
  </View>
  <View
    appearance="default"
    level="2"
    style={
      [
        {
          "backgroundColor": "#eaeaea",
        },
        {
          "flex": 7,
          "flexDirection": "column",
        },
      ]
    }
  >
    <View>
      <View
        appearance="default"
        onSelect={[Function]}
        style={
          [
            {
              "backgroundColor": "#FFFFFF",
              "paddingVertical": 4,
            },
            {
              "flexDirection": "row",
            },
            {
              "height": 50,
            },
          ]
        }
      >
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            status="basic"
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                {
                  "color": "#FF6E14",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Start
          </Text>
        </View>
        <View
          accessibilityState={
            {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
              "opacity": 1,
            }
          }
        >
          <Text
            appearance="default"
            category="p1"
            status="basic"
            style={
              [
                {
                  "color": "#000000",
                  "fontFamily": "System",
                  "fontSize": 15,
                  "fontWeight": "400",
                },
                {
                  "color": "#494949",
                  "fontFamily": "System",
                  "fontSize": 14,
                  "fontWeight": "bold",
                  "marginVertical": 2,
                },
              ]
            }
          >
            Ende
          </Text>
        </View>
      </View>
      <View
        collapsable={false}
        onLayout={[Function]}
        positions={2}
        selectedPosition={0}
        style={
          {
            "transform": [
              {
                "translateX": 0,
              },
            ],
            "width": "50%",
          }
        }
      >
        <View
          style={
            [
              {
                "alignSelf": "center",
                "width": "100%",
              },
              {
                "backgroundColor": "#FF6E14",
                "borderRadius": 2,
                "height": 4,
              },
            ]
          }
          testID="indicator body"
        />
      </View>
    </View>
  </View>
  <View
    appearance="default"
    level="2"
    style={
      [
        {
          "backgroundColor": "#eaeaea",
        },
        {
          "alignItems": "center",
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "marginHorizontal": 5,
        },
      ]
    }
  >
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#b8b8b8",
          "borderColor": "#b8b8b8",
          "borderRadius": 4,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 5,
          "minHeight": 40,
          "minWidth": 40,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        appearance="default"
        category="h1"
        status="basic"
        style={
          [
            {
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 36,
              "fontWeight": "800",
            },
            {
              "color": "#202020",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 10,
            },
          ]
        }
      >
        ABBRECHEN
      </Text>
    </View>
    <View
      accessibilityState={
        {
          "busy": undefined,
          "checked": undefined,
          "disabled": undefined,
          "expanded": undefined,
          "selected": undefined,
        }
      }
      accessibilityValue={
        {
          "max": undefined,
          "min": undefined,
          "now": undefined,
          "text": undefined,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        {
          "alignItems": "center",
          "alignSelf": "center",
          "backgroundColor": "#FF6E14",
          "borderColor": "#FF6E14",
          "borderRadius": 4,
          "borderWidth": 1,
          "flex": 1,
          "flexDirection": "row",
          "justifyContent": "center",
          "margin": 5,
          "minHeight": 40,
          "minWidth": 40,
          "opacity": 1,
          "paddingHorizontal": 10,
          "paddingVertical": 12,
        }
      }
    >
      <Text
        appearance="default"
        category="h1"
        status="basic"
        style={
          [
            {
              "color": "#000000",
              "fontFamily": "System",
              "fontSize": 36,
              "fontWeight": "800",
            },
            {
              "color": "#FFFFFF",
              "fontFamily": "System",
              "fontSize": 14,
              "fontWeight": "bold",
              "marginHorizontal": 10,
            },
          ]
        }
      >
        ERSTELLEN
      </Text>
    </View>
  </View>
</View>
`;
